src/tests/copy.rs

```diff
@@ -185,6 +186,111 @@ fn copy_test() {
+                                     }
+                                 },
+                                 None::<fn(_)>,
+                                 RefCell::new(Vec::<String>::new()));
+        }
+
+        fn get_data_device(&mut self,
+                           _manager: ServerManager,
+                           id: NewResource<ServerDevice>,
+                           _seat: ServerSeat) {
+            let selection = self.selection.clone();
+            id.implement_closure(move |request, _| {
+                                     if let ServerDeviceRequest::SetSelection { source } = request {
+                                         *selection.borrow_mut() = source;
+                                     }
+                                 },
+                                 None::<fn(_)>,
+                                 ());
+        }
+    }
+
+    let mut server = TestServer::new();
+    server.display
+          .create_global::<ServerSeat, _>(6, |new_res, _| {
+              new_res.implement_dummy();
+          });
+
+    let selection = Rc::new(RefCell::new(None));
+    {
+        let selection = selection.clone();
+        server.display
+              .create_global::<ServerManager, _>(1, move |new_res, _| {
+                  new_res.implement(ServerManagerHandler { selection: selection.clone() },
+                                    None::<fn(_)>,
+                                    ());
+              });
+    }
+
+    let socket_name = mem::replace(&mut server.socket_name, OsString::new());
+    let child = thread::spawn(move || {
+        let mut opts = Options::new();
+        opts.foreground(true);
+        let sources = vec![MimeSource { source: Source::Bytes(&[1, 3, 3, 7]), mime: MimeType::Specific("test".to_string()) }, MimeSource { source: Source::Bytes(&[2, 4, 4]), mime: MimeType::Specific("test2".to_string()) }];
+        copy_internal(opts,
+                      sources,
+                      Some(socket_name),
+                      true)
+    });
+
+    thread::sleep(Duration::from_millis(100));
+    server.answer();
+
+    thread::sleep(Duration::from_millis(100));
+    server.answer();
+
+    thread::sleep(Duration::from_millis(100));
+    server.answer();
+
+    let mime_types = selection.borrow().as_ref().map(|x| {
+                                                    x.as_ref()
+                                                     .user_data::<RefCell<Vec<String>>>()
+                                                     .unwrap()
+                                                     .borrow()
+                                                     .clone()
+                                                });
+
+    let (mut read, write) = pipe().unwrap();
+
+    if let Some(source) = selection.borrow().as_ref() {
+        source.send("test".to_string(), write.as_raw_fd());
+        source.send("test2".to_string(), write.as_raw_fd());
```

Comment:
I think you should request these to two separate pipes and check them separately. This will both make for a better test and fix the travis failures.

Reply:
Should be fixed, if I understood correctly